Per-repo config files (and reloading of edited config files)

stefanhaller opened the pull request:
- **PR Description**

Support per-repo user config files. For now we only support `.git/lazygit.yml`; in the future we would also like to support `./.lazygit.yml`, but that one will need a trust prompt as it could be versioned, which adds quite a bit of complexity, so we leave that for later.

We do, however, support config files in parent directories (all the way up to the root directory). This makes it possible to add a config file that applies to multiple repos at once. Useful if you want to set different options for all your work repos vs. all your open-source repos, for instance.

In addition, we support re-loading edited config files. This makes it much easier to experiment with config settings, especially the ones that affect the layout or color scheme, because you see the effect immediately without having to restart lazygit.

- **Please check if the PR fulfills these requirements**

* [x] Cheatsheets are up-to-date (run `go generate ./...`)
* [x] Code has been formatted (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#code-formatting))
* [ ] Tests have been added/updated (see [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/integration/README.md) for the integration test guide)
* [ ] Text is internationalised (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#internationalisation))
* [x] Docs have been updated if necessary
* [x] You've read through your own file changes for silly mistakes etc

stefanhaller commented:
I added two commits that make keybindings and customCommands hot-reloadable, which gets us a lot closer to the desired end state.

jesseduffield reviewed (commented):
Great work, left some comments.

As for aiming for 100% reloadability, I'm not fussed. Config files should rarely change and when they do change, it's pretty easy to reload lazygit. In situations where it's hard to support auto-reload we could instead prompt the user with the option to restart lazygit (without needing to exit it), such that the change takes effect.

stefanhaller commented on docs/dev/Codebase_Guide.md:
```diff
@@ -86,6 +86,12 @@ The event loop is managed in the `MainLoop` function of `vendor/github.com/jesse
 
+## Using UserConfig
+
+The UserConfig struct is loaded from lazygit's global config file (and possibly repo-specific config files). It can be re-loaded while lazygit is running, e.g. when the user edits one of the config files. In this case we should make sure that any new or changed config values take effect immediately. The easiest way to achieve this what we do in most controllers or helpers: these have a pointer to ...
```
You're right. 6a5eb5e0e6d1

stefanhaller commented on pkg/app/entry_point.go:
```diff
@@ -136,6 +136,7 @@ func Start(buildInfo *BuildInfo, integrationTest integrationTypes.IntegrationTes
 
 	if integrationTest != nil {
 		integrationTest.SetupConfig(appConfig)
+		appConfig.SaveGlobalUserConfig()
```
As usual I explained this in the commit message. 😄 

I added a comment to the code in 78d9daf5f10f.

stefanhaller commented on pkg/commands/git_commands/repo_paths.go:
```diff
@@ -64,9 +63,9 @@ func (self *RepoPaths) IsBareRepo() bool {
 	return &RepoPaths{
 		worktreePath:       currentPath,
-		worktreeGitDirPath: path.Join(currentPath, ".git"),
+		worktreeGitDirPath: filepath.Join(currentPath, ".git"),
```
I disagree; we should be using "correct" paths everywhere, and the filepath package is the easy way to achieve that without having to use platform-conditional code everywhere. I noticed this because I tested editing a repo-specific config file on Windows, and upon hitting `e` it showed me a dialog containing `C:\some\repo/.git/lazygit.yml`, which works, but looks very bad.

The only place where we need conditional code is the `repo_paths_test.go` test; that test is now quite awkward I admit, but I still think it's better, because previously the paths used there didn't make any sense on Windows.

jesseduffield replied:
I feel like the blast radius is larger than that: pkg/commands/git_commands/worktree_loader.go hard-codes '/' as a file path separator, as does pkg/gui/presentation/files.go. If we want to be using platform-specific path separators everywhere, we should update those files too, would you agree?

Also, we should keep in mind that we want integration tests to work (some day) on windows, which means being able to force lazygit to use a specific path separator for the sake of keeping conditional logic out of our test assertions.

stefanhaller replied:
> I feel like the blast radius is larger than that: pkg/commands/git_commands/worktree_loader.go hard-codes '/' as a file path separator, as does pkg/gui/presentation/files.go. If we want to be using platform-specific path separators everywhere, we should update those files too, would you agree?

Wow, that makes it trickier. The places that you are pointing out are dealing with relative paths (relative to the repo root), and I'd say those should keep using `/`; even git itself outputs them using `/`, e.g. in `git status -s`. I suppose we should use backslashes only when dealing with absolute paths.

From what I can see, all the places I touched in this commit are dealing with absolute paths, so I'd propose to keep this, and not change the other ones. Does that seem reasonable?

> Also, we should keep in mind that we want integration tests to work (some day) on windows, which means being able to force lazygit to use a specific path separator for the sake of keeping conditional logic out of our test assertions.

I don't see any places in our integration test assertions that expect a path, do you have an example?

I'm more worried about unit tests; it would be bad if we need more ugliness of the kind that I had to add in `repo_paths_test.go` in this commit.

jesseduffield replied:
> From what I can see, all the places I touched in this commit are dealing with absolute paths, so I'd propose to keep this, and not change the other ones. Does that seem reasonable?

Yeah that's reasonable

> I don't see any places in our integration test assertions that expect a path, do you have an example?

I was thinking about any assertions involving the files panel (when not in tree mode). But as you say, those are relative paths so we don't need to worry about them.

> I'm more worried about unit tests; it would be bad if we need more ugliness of the kind that I had to add in repo_paths_test.go in this commit.

Agreed. Though I think if we only encounter such awkwardness with absolute paths, it may not be so bad.

stefanhaller commented on pkg/common/common.go:
```diff
@@ -20,9 +22,9 @@ type Common struct {
 
 func (c *Common) UserConfig() *config.UserConfig {
-	return c.userConfig
+	return c.userConfig.Load().(*config.UserConfig)
```
I was slightly concerned about this too, but I don't think it's an issue. An atomic memory access is not a big deal, and shouldn't make a difference to a normal memory access except in highest concurrency scenarios, which we don't have in lazygit.

While looking at this more, I realized that we don't need `atomic.Value` here; it's enough (and actually even easier) to use an `atomic.Pointer`, which reduces the number of atomic memory accesses from 2 to 1. See a682b8d51b0a.

stefanhaller commented on pkg/config/app_config.go:
```diff
@@ -150,6 +158,10 @@ func loadUserConfig(configFiles []*ConfigFile, base *UserConfig) (*UserConfig, e
 			case ConfigFilePolicyErrorIfMissing:
 				return nil, err
```
Why? `configFile` is not used after this point, I'd actually find this confusing.

jesseduffield replied:
Fair enough

stefanhaller commented:
> In situations where it's hard to support auto-reload we could instead prompt the user with the option to restart lazygit (without needing to exit it), such that the change takes effect.

I thought about that too, I'll consider this as a last resort for cases where it's too much work to do the real reloading (like for the auto-fetch configs). But it might not be trivial to do either; we'll see.

stefanhaller commented:
@jesseduffield I rebased onto master, squashed fixups, and added a few more commits:
- Show a message when changing a config setting that can't be auto-reloaded (2500dfbf9cbe and 60fd9e7f7627)
- Add information about per-repo config files to Config.md (416ba1a9da97)

I consider this ready now.

jesseduffield commented:
@stefanhaller . Noted I'll hopefully get around to reviewing this again soon

jesseduffield commented on pkg/gui/gui.go:
```diff
@@ -434,6 +440,54 @@ func (gui *Gui) onUserConfigLoaded() error {
+		// if the value has changed, ...
+		if !old.Equal(new) {
+			// ... convert the field names to the user-facing names by
+			// lower-casing the first letter, ...
```
This is not guaranteed to work because some fields don't match with their serialized name e.g.:
```go
SuspendOnEdit *bool `yaml:"editInTerminal,omitempty"`
```

Arguably we should just not have mismatched field names, but the reason this happens is that we think of a better name to use internally, but don't want to break backwards compatibility. What are your thoughts on parsing the tag string to pull out the serialized name?

stefanhaller replied:
I was originally planning to read the struct tags, but then was too lazy to figure out how to do that, and hoped I could get away with this simpler version. But you're right, it's not good enough, and actually reading the struct tags is not so hard, see 292e74c4c31c.

jesseduffield reviewed (approved):
Great work! This is a great improvement
